snuba/datasets/errors_replacer.py

```diff
@@ -38,6 +43,66 @@ class ReplacerState(Enum):
     ERRORS = "errors"
 
 
+@dataclass
+class ReplacementContext:
+    all_columns: Sequence[FlattenedColumn]
+    state_name: ReplacerState
+
+
+class Replacement(ReplacementBase):
+    @abstractclassmethod
+    def parse_message(
+        cls, message: Mapping[str, Any], context: ReplacementContext,
+    ) -> Optional["Replacement"]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_excluded_groups(self) -> Sequence[int]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_needs_final(self) -> bool:
+        raise NotImplementedError()
+
+
+EXCLUDE_GROUPS = object()
+NEEDS_FINAL = object()
+QueryTimeFlags = Union[Tuple[object, int], Tuple[object, int, Any]]
+
+# Separate mixin because of https://github.com/python/mypy/issues/5374
```

Comment:
Not sure I get this. 
Legacy Replacement is not abstract, you should be able to make it a dataclass. without having a separate mixin class.
We do it often for Expression https://github.com/getsentry/snuba/blob/master/snuba/query/expressions.py#L116-L117